bdk_core/src/coin_select/coin_selector.rs

```diff
@@ -40,124 +42,85 @@ impl WeightedValue {
-            long_term_feerate: None,
-            min_absolute_fee: 0,
-            base_weight,
-            drain_weight,
-            spend_drain_weight,
-            min_drain_value,
+            feerate: 0.25,
+            min_fee: 0, // TODO figure out what the actual network rule is for this
+            value: 0,
         }
     }
+}
 
-    pub fn fund_outputs(
-        txouts: &[TxOut],
-        drain_output: &TxOut,
-        drain_satisfaction_weight: u32,
-    ) -> Self {
-        let mut tx = Transaction {
-            input: vec![],
-            version: 1,
-            lock_time: LockTime::ZERO.into(),
-            output: txouts.to_vec(),
-        };
-        let base_weight = tx.weight();
-        // this awkward calculation is necessary since TxOut doesn't have \.weight()
-        let drain_weight = {
-            tx.output.push(drain_output.clone());
-            tx.weight() - base_weight
-        };
+impl<'a> CoinSelector<'a> {
+    // TODO: constructor should be number of outputs and output weight instead so we can keep track
+    pub fn new(candidates: &'a [WeightedValue], base_weight: u32) -> Self {
         Self {
-            target_value: if txouts.is_empty() {
-                None
-            } else {
-                Some(txouts.iter().map(|txout| txout.value).sum())
-            },
-            ..Self::from_weights(
-                base_weight as u32,
-                drain_weight as u32,
-                TXIN_BASE_WEIGHT + drain_satisfaction_weight,
-            )
+            base_weight,
+            candidates,
+            selected: Cow::Owned(Default::default()),
+            banned: Cow::Owned(Default::default()),
+            candidate_order: Cow::Owned((0..candidates.len()).collect()),
         }
     }
 
-    pub fn long_term_feerate(&self) -> f32 {
-        self.long_term_feerate.unwrap_or(self.target_feerate)
+    pub fn candidate(&self, index: usize) -> WeightedValue {
+        self.candidates[index]
     }
 
-    pub fn drain_waste(&self) -> i64 {
-        (self.drain_weight as f32 * self.target_feerate
-            + self.spend_drain_weight as f32 * self.long_term_feerate()) as i64
+    pub fn deselect(&mut self, index: usize) -> bool {
+        self.selected.to_mut().remove(&index)
     }
-}
 
-/// [`CoinSelector`] is responsible for selecting and deselecting from a set of canididates.
-#[derive(Debug, Clone)]
-pub struct CoinSelector<'a> {
-    pub opts: &'a CoinSelectorOpt,
-    pub candidates: &'a Vec<WeightedValue>,
-    selected: BTreeSet<usize>,
-}
+    pub fn apply_selection<T>(&self, candidates: &'a [T]) -> impl Iterator<Item = &'a T> + '_ {
+        self.selected.iter().map(|i| &candidates[*i])
+    }
 
-impl<'a> CoinSelector<'a> {
-    pub fn candidate(&self, index: usize) -> &WeightedValue {
-        &self.candidates[index]
+    pub fn select(&mut self, index: usize) -> bool {
+        assert!(index < self.candidates.len());
+        self.selected.to_mut().insert(index)
```

Comment:
I don't know if I am reading wrong, but it seems the `index` here is the `index of the item in the candidate list`?? Where as the value of the stuff in that index is the actual index of the actual utxo (or whatever we are selecting)?? If so, Why keep so many indexes? Cant we just directly select the utxo index in the selected list?